packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -48,14 +53,73 @@ class TimelineData {
   /// The current node in the tree structure of GPU duration events.
   TimelineEvent _gpuEventNode;
 
+  final HeapPriorityQueue<TraceEventWrapper> _cpuTraceHeap =
+      HeapPriorityQueue(traceComparator);
+
+  final HeapPriorityQueue<TraceEventWrapper> _gpuTraceHeap =
+      HeapPriorityQueue(traceComparator);
+
+  static int traceComparator(TraceEventWrapper a, TraceEventWrapper b) {
+    final compare = a.event.timestampMicros.compareTo(b.event.timestampMicros);
+    if (compare == 0 && a.event.name == b.event.name) {
+      // If the events share both their name and their timestamp, ensure the
+      // DurationBegin event has priority over DurationEnd events. Duration
+      // events are the only case where this should matter - we must have a
+      // begin event followed by an end event and we rely on this ordering.
+      return a.event.phase.compareTo(b.event.phase);
+    } else if (compare == 0) {
+      // Otherwise if the events share a timestamp, order them in the order we
```

Comment:
Technically a Comparable function should return 0 if the objects are equal.
What case will that break? 
Potentially returning 0 could cause bugs depending on where the comparable is used.

Reply:
Modified comparator - it should never return 0, but instead default to comparing the events in the order we received them.